Design/spectrum_number_and_workspace_index_abstraction.md

```diff
@@ -0,0 +1,269 @@
+For example, [`LoadEventNexus`, issue 13475](https://github.com/mantidproject/mantid/issues/13475) seems to have an ambiguous interpretation of its inputs.
+
+
+## Internals
+
+#### New functionality
+
+- Introduce a class that deals with translation and validation of indices.
+  This will replace code that is now in `MatrixWorkspace` (translation) and in many algorithms (validation).
+  Let us call this class `Translator` for the rest of this document.
+  - In contrast to the current situation this can be unit tested very well.
+  - Defines a consistent way of dealing with errors in the index specification.
+    This is basically also part of a consistent user interface.
+  - Reduces amount of code in algorithms and makes them more readable.
+- Provide a convenience function or method for declaring all relevant algorithm properties in a single statement.
+  - Reduces amount of code in algorithms and makes them more readable.
+  - Automatically gives consistent interface.
+- Spectrum numbers can be defined differently in each workspace.
+  As a consequence the object dealing with translation must be linked to a workspace.
+  For example, it could be a member of the workspace.
+- We should make sure that we do not negatively influence performance when a trivial set of indices is specified by the user.
+  Trivial can mean either the full range or a range in combination with a trivial mapping from spectrum number to workspace index (such as a simple offset of 1).
+  I think our existing performance tests are in general not sensitive enough for this.
+  However, it will probably be fine if we support some special cases (which are actually the most common ones):
+
+  - We want to avoid going through a `map` or `vector` for translation in the simple cases of "full-range" (all spectra), or a simple "min-max" case.
+    This is simple to do, an example can be found as part of [this pull request](https://github.com/mantidproject/mantid/pull/15465)
+    I believe this will cover 95% of cases and ensure that there will be no performance hit.
+  - In the other cases, when an actual list of spectra is given, I suppose that that list is commonly small (since it is often manual user input), so `map` access is negligible.
+  - The remaining performance critical case is loading event data. Each event has an associated detector ID, and this needs to be translated to a workspace index.
+    This is currently done by creating a `vector` (instead of a `map`) for look-up.
+    We certainly have to support a similar mechanism.
+    This is a case where we will need to write some performance tests, to prevent regression.
+    These tests should be added well in advance so we can gather enough statistics prior to any changes and monitor the introduction of the changes.
+
+To deal with MPI in a way that is hidden from both the user and most developers, `Translator` transforms a user input into, e.g., a `SpectrumIndexSet`.
```

Comment:
@martyngigg Would you also suggest to pick a different name for "detector index", to distinguish it easier from "detector ID"?